src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs

```diff
@@ -39,5 +40,45 @@ public static Span<T> AsSpan<T>(List<T>? list)
         /// <remarks>Items should not be added to or removed from the <see cref="Dictionary{TKey, TValue}"/> while the ref <typeparamref name="TValue"/> is in use.</remarks>
         public static ref TValue? GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key, out bool exists) where TKey : notnull
             => ref Dictionary<TKey, TValue>.CollectionsMarshalHelper.GetValueRefOrAddDefault(dictionary, key, out exists);
+
+        /// <summary>
+        /// Sets the count of the <see cref="List{T}"/> to the specified value.
+        /// </summary>
+        /// <param name="list">The list to set the count of.</param>
+        /// <param name="count">The value to set the list's count to.</param>
+        /// <exception cref="NullReferenceException">
+        /// <paramref name="list"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="count"/> is negative.
+        /// </exception>
+        /// <remarks>
+        /// When increasing the count, uninitialized data is being exposed.
+        /// </remarks>
+        public static void SetCount<T>(List<T> list, int count)
+        {
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(count));
+            }
+
+            if (count == list._size)
+            {
+                return;
+            }
+
+            list._version++;
+
+            if (count > list.Capacity)
+            {
+                list.Grow(count);
+            }
+            else if (count < list._size && RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                Array.Clear(list._items, count, list._size - count);
+            }
+
+            list._size = count;
```

Comment:
If we expect the most common use case will be increasing the count, should we instead reorder the method to something like:
```C#
public static void SetCount<T>(List<T> list, int count)
{
    if (count >= list._size)
    {
        if (count > list._items.Length)
        {
            list.Grow(count);
        }
    }
    else
    {
        if (count < 0)
        {
            ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(count));
        }

        if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
        {
            Array.Clear(list._items, count, list._size - count);
        }
    }

    list._size = count;
    list._version++;
}
```
?  This will still increment the version if the count doesn't change, but I don't think that's much different than `list[i] = list[i]`, which will also increment the version.

Reply:
I feel like this just makes the code harder to read for questionable benefit. Just removing the zero check for my version would be just as efficient while being easier to read. Do we want to do that?

Reply:
The < 0 check is required. It could be in a different place, as I did.

Reply:
> The < 0 check is required. It could be in a different place, as I did.

Sorry, I've meant the `count == list._size` check.

Reply:
@stephentoub I think you missed my reply.

Reply:
> Just removing the zero check for my version would be just as efficient while being easier to read.

How so?  The two common cases we expect are:
- Increase the size without needing to increase the capacity
- Increase the size with needing to increase the capacity

The latter involves a non-inlined method call, allocation, etc., so I'm not focusing on it here.

For the former, with the code as written in this PR, it'll go through four conditions/branches.  If you remove the == check, it'll go through three.  What I suggested has two.

Reply:
@stephentoub I've checked the diffs when increasing the size but not growing, for unmanaged types, your version is generally worse (I assume value types to be the main usecase here).
For ref types your version saves basically a single jump that is always taken, at the cost of bigger overall codegen.
Is pessimizing value types worth the cost here?
For context, I've removed the same size check from current version while comparing.

Another thing is, since both versions aren't inlined, should they be marked with AggresiveInlining?

Reply:
> I've checked the diffs when increasing the size but not growing, for unmanaged types, your version is generally worse (I assume value types to be the main usecase here)

Did you actually benchmark it?  I'm surprised at the conclusion.

That said, it's not worth holding up the PR for.  I'll find time to measure it myself subsequently.

> should they be marked with AggresiveInlining?

I don't think it's worth bloating the call sites.  If we want to restructure it such that the common case is inlineable and everything else is movable to a non-inlined slow path, that's fine.  At this point, though, let's just get it in.